HaveIBeenPwned weak password validation

simeonbizhub opened the pull request:
[Trello](https://trello.com/c/YrsVlD9Y)

**Changes**
- Added `HIBPPasswordValidator.isPasswordPwned` class & method
- Added `Configuration.haveIBeenPwnedEnabled` attr & condition
- Updated all password set/change spots to additionally validate via `HIPPasswordValidator` (if enabled)

**Questions**
1. This extra validation is configurable - should it be enabled or disabled by default?
2. I have the API calls surrounded in a try-catch - if the call fails, I am simply logging and continuing. What would you like me to do in these scenarios?

benpetito commented on skyve-ext/src/main/java/org/skyve/impl/security/HIBPPasswordValidator.java:
```diff
@@ -0,0 +1,81 @@
+	private static boolean isHashSuffixPresent(String response, String suffix) {
+		String[] lines = response.split("\r\n");
+		for (String line : lines) {
+			if (line.startsWith(suffix)) {
```
Why is this doing a `startsWith` check? Isn't it comparing hashes?

simeonbizhub replied:
The API response includes the **{suffix}:{count}** of every hash beginning with the specified prefix (line 37). 
I am using `startsWith` rather than `equals` as the line includes ':{count}' in addition the hash suffix.

See https://haveibeenpwned.com/API/v3#PwnedPasswords

benpetito commented on skyve-war/src/main/java/modules/admin/ChangePassword/actions/MakePasswordChange.java:
```diff
@@ -47,11 +48,28 @@ public ServerSideActionResult<ChangePassword> execute(ChangePassword bean, WebCo
+			} catch (ValidationException ve) {
+				throw ve;
+			} catch (Exception e) {
+				Util.LOGGER.severe("HaveIBeenPwned API call failed");
```
I would change this to a `warning`, not `severe`, as it does not impact the overall health of the system.

benpetito commented on skyve-war/src/main/java/modules/admin/ChangePassword/actions/MakePasswordChange.java:
```diff
@@ -47,11 +48,28 @@ public ServerSideActionResult<ChangePassword> execute(ChangePassword bean, WebCo
+			try {
+				// validate if password is PWNED
+				if (HIBPPasswordValidator.isPasswordPwned(newPassword)) {
+					Message message = new Message(ChangePassword.newPasswordPropertyName,
+							"The password you have entered has been breached. "
+									+ "Please re-enter and confirm the password.");
+
+					throw new ValidationException(message);
```
Is there a way we can implement this as a warning as per the Trello ticket? If this is enabled, this will prevent someone from using a pwned password at all won't it?

benpetito commented on skyve-war/src/main/java/modules/admin/ChangePassword/actions/MakePasswordChange.java:
```diff
@@ -47,11 +48,28 @@ public ServerSideActionResult<ChangePassword> execute(ChangePassword bean, WebCo
 			Message message = new Message(ChangePassword.newPasswordPropertyName,
 					"You did not type the same password.  Please re-enter and confirm the password.");
 			message.addBinding(ChangePassword.confirmPasswordPropertyName);
-			throw new ValidationException(message);
```
Was this supposed to be removed? Don't we still want this check?

benpetito commented on skyve-war/src/main/java/modules/admin/Configuration/Configuration.xml:
```diff
@@ -43,6 +43,12 @@
 		</text>
+		<boolean name="haveIBeenPwnedEnabled">
+			<documentation>Determines whether or not HaveIBeenPwned API is used as part of password validation.</documentation>
+			<displayName>HaveIBeenPwned API enabled</displayName>
+			<description>The HaveIBeenPwned API checks whether a password has been compromised in known data breaches.</description>
```
I wonder if we should change the terminology for the end user to be more generic, as it is unlikely they know what HaveIBeenPwned is.

Something along the lines of:

Check for breached password
When users try to create or change a password, checks whether the new password has been compromised in known data breaches (requires internet access).

benpetito commented on skyve-war/src/main/java/modules/admin/SelfRegistration/SelfRegistrationExtension.java:
```diff
@@ -109,6 +112,22 @@ public void validateConfirmPassword() {
+
+					}
+				} catch (Exception e) {
+					Util.LOGGER.severe("HaveIBeenPwned API call failed");
```
Change to warning.

benpetito commented on skyve-war/src/main/java/modules/admin/User/UserBizlet.java:
```diff
@@ -300,7 +302,23 @@ public static void validateUserNameAndPassword(UserExtension user, ValidationExc
+
+								}
+							} catch (Exception e) {
+								Util.LOGGER.severe("HaveIBeenPwned API call failed");
```
Change to warning.

benpetito commented:
Hi @simeonbizhub,

I think this should be on by default, otherwise it will never get used.

I think just logging the exception is fine, just change the log level. Severe is typically reserved for things that will impact the system staying up without a restart or requiring direct intervention.

If possible, I would prefer this to just be a warning to the user, instead of a validation exception which prevents them from using a breached password. I'm not sure if we can use an EXT.pushMessage or a condition to show a warning or something?

simeonbizhub commented:
Thanks @benpetito,

I have applied the changes requested, including altering the implementation to simply warn the user they have selected a breached password, rather than throwing a validation exception (growl with message severity = warning).

I achieved this by evaluating the new password in `rerender` of select `bizlets` - and this works - except for the password reset change password view (`changePassword.jsp`). 

Do you have any idea of what I can do here to achieve the same result as in the rest of the views? 
Or should I be taking a different approach?

benpetito commented on skyve-war/src/main/java/modules/admin/User/UserBizlet.java:
```diff
@@ -94,6 +94,19 @@ public void preRerender(String source, UserExtension bean, WebContext webContext
+				if (c.isCheckForBreachedPasswordsEnabled()) {
+					if (HIBPPasswordValidator.isPasswordPwned(newPassword)) {
+						webContext.growl(MessageSeverity.warn,
+								"WARNING: The password you have entered has been compromised in a data breach. For security, we recommend selecting a stronger password.");
```
Because this message is used in a couple of places, I would probably externalise the string. Makes it easier to update the warning in one place (and translate).

benpetito commented on skyve-ext/src/main/java/org/skyve/impl/security/HIBPPasswordValidator.java:
```diff
@@ -18,34 +22,41 @@ public class HIBPPasswordValidator {
-				.uri(URI.create(HIBP_API_URL + prefix))
-				.GET()
-				.build();
+			// ]Query the HaveIBeenPwned API with the prefix
```
I think you have an extra `]` here by mistake

benpetito reviewed (changes requested):
Hey @simeonbizhub , just a couple of minor tweaks sorry.

simeonbizhub commented:
> Hey @simeonbizhub , just a couple of minor tweaks sorry.

Thanks @benpetito, change requests resolved.

simeonbizhub commented:
Hey @benpetito , I implemented a 2-stage submit for changePassword.jsp & passwordReset.jsp, as suggested.

Let me know your thoughts. Thanks

simeonbizhub commented:
Hey @benpetito , I added the 'is check for breach passwords enabled' to changePassword.jsp & resetPassword.jsp.

To achieve this, I moved checkForBreachedPassword from Configuration.xml to Startup.xml, utilising the override json.

Let me know of your feedback - thanks.

simeonbizhub commented:
@benpetito I have updated changePassword.jsp & resetPassword.jsp to additionally store the hashed password in the session, so that we can evaluate if the warning has been displayed for that exact password.

benpetito commented:
Hi @simeonbizhub , this doesn't seem to work. I can never change my password. It successfully warned me with different breached passwords, but I was unable to enter it twice in a row and actually change it. I was stuck in a loop.

matforest commented:
Hey, the behaviour of [this onChange/rerender](https://github.com/skyvers/skyve/blob/3fa819c6d09aba95bfc83d0cd7bc9f1ed6f50f8c/skyve-war/src/main/java/modules/admin/User/views/edit.xml#L89-L93) is a bit weird in smart client; after each key stroke typing is interrupted (focus is lost for a moment) which makes typing normally impossible. Would an onBlur handler work better here instead of onChange?
